File: scripts/traktor.py
import socket
import re
import sys
from os import path
from struct import unpack

filename = path.realpath(path.dirname(sys.argv[0])) + '/' + 'NowPlaying.txt'
meta = { 'title': None, 'artist': None }
data = ''

def saveData():
    global filename
    global meta
    metaData = "Title: " + meta['title'] + "\nArtist: " + meta['artist']
    datafile = open(filename, 'w')
    datafile.write(metaData)
    datafile.close()

def getName(needle, haystack):
    regmatch = '(....)' + needle.upper() + '=(.*)'
    name = re.search(regmatch.encode('utf-8'), haystack)
    if name:
        length = unpack('=l', name.group(1))[0]
        length = length - len(needle) - 1
        if len(name.group(2)) >= length:
            return (name.group(2))[0:length]
    return None

sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.bind(('localhost', 8000))
sock.listen(1)
conn, addr = sock.accept()
print 'Connected:', addr

while True:
    dataNew = conn.recv(4096)
    if not dataNew:
        break
    data += dataNew
    if data.find('ice') != -1:
        conn.sendall("HTTP/1.0 200 OK\r\n\r\n")
        data = ''
        continue
    meta['artist'] = getName('artist', data)
    meta['title'] = getName('title', data)
    if meta['artist'] is not None and meta['title'] is not None:
        saveData()
        print meta['artist'], "-", meta['title']
        meta['artist'] = None
        meta['title'] = None
        dataNew = ''
    data = dataNew

conn.close()

File: scripts/traktor.js
const fs = require('fs');
const net = require('net');
const bufferpack = require('./bufferpack');

var needInit = true;
var datafile = __dirname + '/' + 'NowPlaying.txt';

net.createServer(function (sock) {
	sock.on('data', function (dataChunk) {
		if (needInit) {
			sock.write('HTTP/1.0 200 OK\r\n\r\n');
			console.log('Conected');
			needInit = false;
		} else {
			var title = getName('title', dataChunk);
			var artist = getName('artist', dataChunk);
			if (title || artist) {
				fs.writeFile(datafile, 'Title: ' + (title || '') + '\nArtist: ' + (artist || ''), function () {
					console.log(artist + ' - ' + title);
				});
			}
		}
	});
	sock.on('end', function () {
		needInit = true;
		console.log('Disconected. Ready to connect');
	});
}).listen(8000, 'localhost', function () {
	console.log('Ready to connect\nCtrl+C to exit...');
});

/**
 * Search in Buffer
 */
function getName(needle, haystack) {
	// (....)TITLE=(.*) - $1: размер, $2: значение
	needle = needle.toUpperCase() + '=';
	var position = haystack.indexOf(needle);
	if (position != -1) {
		var length = haystack.slice(position - 4, position);
		length = bufferpack.unpack('<l', length);
		return haystack.toString('utf8', position + needle.length, position + length[0]);
	}
}
